apps/labs/posts/scipy-array-api.md

```diff
@@ -0,0 +1,441 @@
+[the Python array API standard](https://data-apis.org/array-api/), which is a "specification for what it means for a
+library to have an ‘array API’" (see [Aaron's talk](https://youtu.be/16rB-fosAWw)).
+Array consumer libraries can write 'array-agnostic' code which is compatible with this specification,
+rather than any particular provider library's API, and it will work with any provider library which complies with the
+standard.
+NumPy, CuPy and PyTorch are all planning to be fully compliant - with this being one of
+[the goals for NumPy 2.0](https://github.com/orgs/numpy/projects/9) - and JAX and Dask also have implementations
+in progress.
+The consortium has also developed [array-api-compat](https://github.com/data-apis/array-api-compat),
+a small wrapper around each array library (currently NumPy, CuPy and PyTorch), which, for practical purposes,
+is fully compliant with the standard.
+
+This means that array consumer libraries are able to start writing and testing array-agnostic code today.
+For a NumPy-consuming library like SciPy, this has opened the opportunity to implement support for CuPy and
+PyTorch by adding support for the standard into submodules.
+
+## SciPy Support for the Array API
+
+Before these interoperability efforts started, SciPy did not work well with provider libraries other than
+NumPy at all (as expressed by
+[this amazing illustration](https://labs.quansight.org/posts/array-libraries-interoperability/array_wonderland.JPG)
+from [Anirudh's blog post](https://labs.quansight.org/blog/array-libraries-interoperability)).
+For example, when trying to use `scipy.fft.fft` with a CuPy array, users would be hit with this error message:
+
+```python
+>>> import scipy
+>>> import numpy as np
+>>> scipy.fft.fft(np.exp(2j * np.pi * np.arange(8) / 8))
+array([-3.44509285e-16+1.14423775e-17j,
+       …
+       0.00000000e+00+1.22464680e-16j])
+
+>>> import cupy as cp
+>>> scipy.fft.fft(cp.exp(2j * cp.pi * cp.arange(8) / 8))
+…
+KeyError: 'ALIGNED is not defined for cupy.ndarray.flags'
+```
+
+Adopting the standard is the first step towards a future where users who have programs made up of
+SciPy functions can simply change the provider library used for their input data, and have the rest of the
+program still work the same, while taking advantage of any benefits provided by the new provider library.
+
+To kick off the process of this work, the request-for-comment (RFC) issue
+["SciPy array types & libraries support"](https://github.com/scipy/scipy/issues/18286) was opened.
+This set out the basic design principle of "***container type in == container type out***", as well as a
+detailed plan for how to treat different array inputs, and the development strategy.
+To summarise, the scope of the work is to treat all array inputs as specified in the RFC and to convert all
+existing pure Python + NumPy code to be array API compatible.
+Out of scope is converting the C/C++/Cython/Fortran code which is used within SciPy.
```

Comment:
You explain this a bit bellow (`fht` section), but here it might not be clear why. If readers do not go read on the other posts, they would not really get the subtlety.

I suggest to add a short 1-2 lines to explain this. You could also add that behind the scene SciPy provides a Python API for low level functions written in other languages. If you say that, it would probably fit better where you describe first SciPy and you would emphasis the fact that these implementations in other languages are CPU only.

Reply:
```suggestion
Out of scope is converting the C/C++/Cython/Fortran code which is used within SciPy.
Behind the scenes, SciPy provides Python APIs for low level functions which are written in other languages and for the
CPU only.
These APIs are made to work with NumPy arrays - they will definitely not work for arrays on a different device to the CPU.
For these functions which use compiled code, some dispatching mechanism...
```

Reply:
I went with "Behind the scenes, SciPy provides Python APIs for low level functions which are written in other languages and for the CPU only. These APIs are made to work with NumPy arrays - they will definitely not work for arrays on a different device to the CPU, and are not guaranteed to work with other types of array on the CPU. For these functions which use compiled code, some dispatching mechanism seems like the right solution, but no particular implementation or development strategy has been settled on yet - see the discussion of `uarray` in [Anirudh's blog post](https://labs.quansight.org/blog/array-libraries-interoperability) and [the RFC](https://github.com/scipy/scipy/issues/18286) for more information."